### PR TITLE
Port definitions refactor

### PR DESCRIPTION
**Pull request by tkornuta-nvidia**

### CODE REVIEW

**Comment by tkornuta-nvidia** [comment]
in file pytorch/common/others.py:
class SequenceProjection(TrainableNM):

    def input_port_definitions(self):
        """Returns definitions of module input ports.

        input_seq:
            Empty Type?!?
        """
        return {
            "input_seq": NeuralType({})
        }

**Comment by tkornuta-nvidia** [comment]
What are the outputs of WaveGlowNM?


    def output_port_definitions(self):
        """Returns definitions of module output ports.

        audio:
            0: AxisType(BatchTag)

            1: AxisType(TimeTag)

        log_s_list:
            List?

        log_det_W_list:
            List?
        """
        return {
            "audio": NeuralType({
                0: AxisType(BatchTag),
                1: AxisType(TimeTag)}),
            "log_s_list": NeuralType(),
            "log_det_W_list": NeuralType(),
        }

**Comment by tkornuta-nvidia** [comment]
@redoctopus need some help with mandarin documentation, in custommodules.rst there is reference to create_port() whic is obsolete...

**Comment by tkornuta-nvidia** [comment]
*Info*: this PR removes deprecated method of passing some instance-dependent arguments to static method by using *create_port_args*

**Comment by tkornuta-nvidia** [comment]
In export had to comment some of the "custom hacks" that were removing some of the input and output ports from JasperEncoder class:

        # Custom hacks
        # We are checking type like this to avoid taking dependency on nemo_asr
        if type(module).__name__ == "JasperEncoder":
            # TK: Commenting this as this is not the way it works now.
            # Not sure what are the consequences...

            # self.logger.warning(f"Module is JasperEncoder. We are removing"
            #                    f"input and output length ports since they "
            #                    f"are not needed for deployment")
            #del module._input_ports['length']
            #del module._output_ports['encoded_lengths']

@okuchaiev please take a look at that...

**Comment by tkornuta-nvidia** [comment]
After the fix the unit tests seem to be working correctly, but one thing that bothers me is NaN loss in one of the unit tests (before the Parallel stage):

2019-12-19 21:08:41,116 - INFO - Starting epoch 0
WARNING:root:WARNING: Loss is NaN
2019-12-19 21:08:41,320 - WARNING - WARNING: Loss is NaN
WARNING:root:WARNING: Loss is NaN
2019-12-19 21:08:41,528 - WARNING - WARNING: Loss is NaN
WARNING:root:WARNING: Loss is NaN
2019-12-19 21:08:41,737 - WARNING - WARNING: Loss is NaN
WARNING:root:WARNING: Loss is NaN

@okuchaiev is that ok?

File: nemo/nemo/backends/pytorch/module_wrapper.py [hunk 1]
```diff
@@ -10,13 +10,6 @@ class TrainableNeuralModuleWrapper(NeuralModule, nn.Module):
     """This class wraps an instance of Pytorch's nn.Module and
     returns NeuralModule's instance."""
 
-    @staticmethod
```
Comment (blisc) [inline nemo/nemo/backends/pytorch/module_wrapper.py]: Can we throw a deprecation error instead of removing this function?

File: nemo/nemo/core/neural_modules.py [hunk 2]
```diff
@@ -89,6 +82,22 @@ def __init__(
         #        "arguments:".format(self.__class__.__name__))
         #    self._logger.warning("{}".format(kwargs.keys()))
 
+    @abstractmethod
+    def input_port_definitions(self) -> Optional[Dict[str, NeuralType]]:
```
Comment (blisc) [inline nemo/nemo/core/neural_modules.py]: I remember talking abut this, but can you remind me why this is a method instead of a property?

**Comment by blisc** [comment]
> What are the outputs of WaveGlowNM?
> 
> ```
> def output_port_definitions(self):
>     """Returns definitions of module output ports.
> 
>     audio:
>         0: AxisType(BatchTag)
> 
>         1: AxisType(TimeTag)
> 
>     log_s_list:
>         List?
> 
>     log_det_W_list:
>         List?
>     """
>     return {
>         "audio": NeuralType({
>             0: AxisType(BatchTag),
>             1: AxisType(TimeTag)}),
>         "log_s_list": NeuralType(),
>         "log_det_W_list": NeuralType(),
>     }
> ```

Can you add a TODO comment tagging me?

File: nemo/nemo/core/neural_factory.py [hunk 3]
```diff
@@ -651,11 +651,14 @@ def deployment_export(self,
         # Custom hacks
         # We are checking type like this to avoid taking dependency on nemo_asr
         if type(module).__name__ == "JasperEncoder":
-            self.logger.warning(f"Module is JasperEncoder. We are removing"
-                                f"input and output length ports since they "
-                                f"are not needed for deployment")
-            del module._input_ports['length']
-            del module._output_ports['encoded_lengths']
+            # TK: Commenting this as this is not the way it works now.
```
Comment (okuchaiev) [inline nemo/nemo/core/neural_factory.py]: please put this back - your change will break NeMo->Jarvis export for Jasper/QuartzNet

File: nemo/nemo/core/neural_modules.py [hunk 4]
```diff
@@ -115,15 +124,19 @@ def __call__(self, **kwargs):
         #    raise ValueError("We currently do not support calling same NM"
         #                     "more than once")
 
+        # Get input and output ports definitions.
+        input_port_defs = self.input_port_definitions()
+        output_port_defs = self.output_port_definitions()
```
Comment (okuchaiev) [inline nemo/nemo/core/neural_modules.py]: why don't use "self.input_port_definitions()" and "self.output_port_definitions()" directly?

File: nemo/nemo/core/neural_modules.py (hunk 2 above)
Comment (okuchaiev) [reply nemo/nemo/core/neural_modules.py]: Delete  "input_port_definitions" and "output_port_definitions". You can keep using "input_ports" and "output_ports" properties instead. 
This will require much fewer changes throughout the codebase and will make the transition easier.

File: nemo/nemo/core/neural_modules.py [hunk 5]
```diff
@@ -115,15 +124,19 @@ def __call__(self, **kwargs):
...
+
         first_input_nmtensor_type = None
         input_nmtensors_are_of_same_type = True
         for port_name, tgv in kwargs.items():
-            if port_name not in self._input_ports.keys():
+            if port_name not in input_port_defs.keys():
```
Comment (okuchaiev) [inline nemo/nemo/core/neural_modules.py]: basically, if you address my comment above, changes like these are unnecessary

File: nemo/nemo/core/neural_modules.py [hunk 6]
```diff
@@ -363,24 +376,6 @@ def local_parameters(self) -> Optional[Dict]:
         """
         return self._local_parameters
 
-    @property
-    def input_ports(self) -> Optional[Dict[str, NeuralType]]:
```
Comment (okuchaiev) [inline nemo/nemo/core/neural_modules.py]: keep these instead of your newly introduced input_port_definitions and output_port_definitions

File: nemo/nemo/core/neural_modules.py [hunk 7]
```diff
@@ -404,19 +399,6 @@ def num_weights(self):
         """
         pass
 
-    @staticmethod
-    @abstractmethod
-    def create_ports(**kwargs):
-        """
```
Comment (okuchaiev) [inline nemo/nemo/core/neural_modules.py]: do not delete this method. 
Here should come deprecation warning saying how things should be done now. It also should keep behaving as it does now (but with deprecation warning). This will be in +1 release. In +2 release this method still should be there but will be throwing an exception saying how things should be done now

File: nemo/nemo/core/neural_factory.py (hunk 3 above)
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_factory.py]: I cannot simply "put it back", as there are no _input_ports and _output_ports attributes of the module that you can delete from. 

My proposition is to retrieve those objects using provided methods and _then_ remove the ports you don't need.

Let's meet and discuss that.

File: nemo/nemo/core/neural_modules.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_modules.py]: Hey, so this is "just a matter of taste."
https://stackoverflow.com/questions/20506612/python-property-vs-method-when-no-access-to-attribute-is-needed

IMO in several cases we are performing some actual calculations there (e.g. generating list of ports etc.), so I decided that those getters should be methos.

But if majority of you want it to be changed into property, I have no problem with
replacing @abstractmethod with @abstractproperty ;)

File: nemo/nemo/core/neural_modules.py (hunk 4 above)
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_modules.py]: less changes in code ;)

File: nemo/nemo/core/neural_modules.py (hunk 5 above)
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_modules.py]: I find those names more appropriate - we are talking about _port definitions_ not _ports_ per se.

File: nemo/nemo/core/neural_modules.py (hunk 6 above)
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_modules.py]: Please find my comment above.

File: nemo/nemo/core/neural_modules.py (hunk 7 above)
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_modules.py]: I have already changed the documentation and explained how one should use those two methods instead of deprecated create_ports.

Of course we can leave create_ports (not abstract, but simply returning the deprecation warning), but please note that one still will have to implement _input_port_definitions()_ and _output_port_definitions()_, without them the module will be abstract - so what is really the point here, for me leaving that leaves a mess, not backward compatibility...

File: nemo/nemo/core/neural_modules.py (hunk 2 above)
Comment (okuchaiev) [reply nemo/nemo/core/neural_modules.py]: it is not a matter of taste. If it is a property one could use "myclass.input_ports" not "myclass.input_ports()". We aren't going to pass any args there, so it should be property then.

File: nemo/nemo/core/neural_factory.py (hunk 3 above)
Comment (okuchaiev) [reply nemo/nemo/core/neural_factory.py]: why remove _input_ports and _output_ports back though?

File: nemo/nemo/core/neural_modules.py (hunk 5 above)
Comment (okuchaiev) [reply nemo/nemo/core/neural_modules.py]: Keeping old names keeps more backward of the compatibility intact and is less error prone.
And if we want to debate over the name (which I don't think we should because backwards compatibility precedes that) I find both "input_port_definitions" and "input_ports" too long, should be just changed to "inputs"

File: nemo/nemo/core/neural_modules.py (hunk 7 above)
Comment (okuchaiev) [reply nemo/nemo/core/neural_modules.py]: I thought we were basically changing from static "create_ports" to member method "create_ports" (or member method "create_input_ports" and "create_output_ports").
That can be done with minimal, if any, breaking changes. Am I missing something?

Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_modules.py]: 1) There are no braking changes.
2) I have always said that "create" is misleading, because we are not creating any physical instances of ports in those methods (btw. please note that you want those to be properties, not methods).

File: nemo/nemo/core/neural_modules.py (hunk 5 above)
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_modules.py]: 1) we are discussing changing of *two lines of code inside of neural_module*
2) I guess we both agree on naming convention that *underscore* at the beginning shouldn't be used for names of externally visible methods/properties. So the name _input_ports cannot remain as it is, am I correct?

File: nemo/nemo/core/neural_factory.py (hunk 3 above)
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_factory.py]: Because those were *structures* that we don't need at all. The idea was that the methods/properties will return definitions on the fly, when called.

We have spent some significant time on those discussions and I thought we agreed on that.

**Comment by tkornuta-nvidia** [comment]
> The changes generally look good to me. I am leaving @okuchaiev with final approval.



> > What are the outputs of WaveGlowNM?
> > ```
> > def output_port_definitions(self):
> >     """Returns definitions of module output ports.
> > 
> >     audio:
> >         0: AxisType(BatchTag)
> > 
> >         1: AxisType(TimeTag)
> > 
> >     log_s_list:
> >         List?
> > 
> >     log_det_W_list:
> >         List?
> >     """
> >     return {
> >         "audio": NeuralType({
> >             0: AxisType(BatchTag),
> >             1: AxisType(TimeTag)}),
> >         "log_s_list": NeuralType(),
> >         "log_det_W_list": NeuralType(),
> >     }
> > ```
> Can you add a TODO comment tagging me?

Ok, doing that. But generally - are those lists of any kind?

**Comment by okuchaiev** [comment]
@tkornuta-nvidia let me push export workaround first

File: nemo/nemo/backends/pytorch/torchvision/data/image_folder.py [hunk 8]
```diff
@@ -9,10 +9,24 @@ class ImageFolderDataLayer(DataLayerNM):
...
+            2: AxisType(HeightTag, input_size)
+
+            3: AxisType(WidthTag, input_size)
+
+
+        label:
+            0: AxisType(BatchTag)
+        """
+        return {
             "image": NeuralType(
```
[the remark this answers is not in the document]
Comment (tkornuta-nvidia) [reply nemo/nemo/backends/pytorch/torchvision/data/image_folder.py]: ok, fixing that

File: nemo/nemo/core/neural_modules.py [hunk 9]
```diff
@@ -89,6 +84,31 @@ def __init__(
         #        "arguments:".format(self.__class__.__name__))
         #    self._logger.warning("{}".format(kwargs.keys()))
 
+    @staticmethod
+    def create_ports(**kwargs):
+        """ Deprecated method, to be remoted in the next release."""
+        warnings.warn(
```
[the remark this answers is not in the document]
Comment (tkornuta-nvidia) [reply nemo/nemo/core/neural_modules.py]: done

**Comment by tkornuta-nvidia** [comment]
ok, removing the branch
